File: pcc/py_runtime/py/py_weakref.py
"""Minimal pcc-Python port of weakref.ref runtime support."""

from pcc.extern import extern, c_abi_export, c_ptr, c_int32, c_int64, c_void
from pcc.unsafe import (
    calloc,
    cstr,
    free,
    global_load_ptr,
    global_store_ptr,
    is_tagged_int,
    load_i32,
    load_ptr,
    null,
    ptr_add,
    ptr_eq,
    ptr_is_null,
    store_i32,
    store_i64,
    store_ptr,
)


py_incref = extern("py_incref", (c_ptr,), c_void)
py_decref = extern("py_decref", (c_ptr,), c_void)
py_dict_del = extern("py_dict_del", (c_ptr, c_ptr), c_int64)
py_dict_get = extern("py_dict_get", (c_ptr, c_ptr), c_ptr)
py_dict_keys = extern("py_dict_keys", (c_ptr,), c_ptr)
py_dict_new = extern("py_dict_new", (), c_ptr)
py_dict_set = extern("py_dict_set", (c_ptr, c_ptr, c_ptr), c_void)
py_list_append = extern("py_list_append", (c_ptr, c_ptr), c_void)
py_list_get = extern("py_list_get", (c_ptr, c_int64), c_ptr)
py_list_len = extern("py_list_len", (c_ptr,), c_int64)
py_list_new = extern("py_list_new", (c_int64,), c_ptr)
py_list_pop = extern("py_list_pop", (c_ptr, c_int64), c_ptr)
py_list_set = extern("py_list_set", (c_ptr, c_int64, c_ptr), c_void)
py_tuple_new = extern("py_tuple_new", (c_int64,), c_ptr)
py_tuple_get = extern("py_tuple_get", (c_ptr, c_int64), c_ptr)
py_tuple_set_item = extern("py_tuple_set_item", (c_ptr, c_int64, c_ptr), c_void)
py_obj_call = extern("py_obj_call", (c_ptr, c_ptr, c_ptr), c_ptr)
py_clear_exception = extern("py_clear_exception", (), c_void)
py_exc_new = extern("py_exc_new", (c_int64, c_ptr), c_ptr)
py_raise = extern("py_raise", (c_ptr,), c_void)
pcc_runtime_log_event_code = extern("pcc_runtime_log_event_code", (c_int32, c_int32, c_int64, c_int64, c_ptr), c_void)
pcc_gc_alloc = extern("pcc_gc_alloc", (c_int64, c_int32, c_int32), c_ptr)
pcc_gc_pin = extern("pcc_gc_pin", (c_ptr,), c_void)
pcc_gc_note_relocation_read = extern(
    "pcc_gc_note_relocation_read", (c_ptr,), c_ptr,
)
pcc_gc_load_ptr = extern("pcc_gc_load_ptr", (c_ptr, c_ptr), c_ptr)
pcc_gc_store_ptr = extern("pcc_gc_store_ptr", (c_ptr, c_ptr, c_ptr), c_void)
pcc_gc_free_object_memory = extern("pcc_gc_free_object_memory", (c_ptr,), c_void)


def _py_none():
    return global_load_ptr("py_None")


def _unlink(wr) -> None:
    prev = load_ptr(wr, 32)
    nxt = load_ptr(wr, 40)
    if ptr_eq(prev, wr) != 0:
        store_ptr(wr, 32, null())
        store_ptr(wr, 40, null())
        return
    if ptr_is_null(prev) != 0:
        if ptr_eq(global_load_ptr("py_weakref_head"), wr) == 0:
            store_ptr(wr, 40, null())
            return
    else:
        if ptr_eq(load_ptr(prev, 40), wr) == 0:
            store_ptr(wr, 32, null())
            store_ptr(wr, 40, null())
            return
    if ptr_is_null(prev) != 0:
        global_store_ptr("py_weakref_head", nxt)
    else:
        store_ptr(prev, 40, nxt)
    if ptr_is_null(nxt) == 0:
        store_ptr(nxt, 32, prev)
    store_ptr(wr, 32, null())
    store_ptr(wr, 40, null())


@c_abi_export("py_weakref_new")
def py_weakref_new(target, callback):
    if ptr_is_null(target) != 0:
        return null()
    if is_tagged_int(target) != 0:
        return null()
    # Value projections are identity-free; a ValueBox's lifetime is
    # unpredictable (every boxing makes a new box) — mirror CPython's
    # analogue (weakref.ref(3) -> TypeError). 200 == PY_TYPE_VALUEBOX,
    # 3 == PY_EXC_TYPEERROR. The compile-time diagnostic catches the
    # static form; this covers Dyn-path boxes.
    if load_i32(target, 8) == 200:
        py_raise(
            py_exc_new(
                3,
                cstr("cannot create weak reference to a valueclass payload"),
            )
        )
        return null()
    if ptr_eq(callback, _py_none()) != 0:
        callback = null()
    target = pcc_gc_note_relocation_read(target)

    wr = pcc_gc_alloc(48, 21, 0)
    if ptr_is_null(wr) != 0:
        return null()
    store_i64(wr, 0, 1)
    store_i32(wr, 8, 21)        # PY_TYPE_WEAKREF
    store_ptr(wr, 16, target)
    store_ptr(wr, 24, null())
    store_ptr(wr, 32, null())
    store_ptr(wr, 40, null())
    if ptr_is_null(callback) == 0:
        pcc_gc_store_ptr(wr, ptr_add(wr, 24), callback)

    head = global_load_ptr("py_weakref_head")
    store_ptr(wr, 40, head)
    if ptr_is_null(head) == 0:
        store_ptr(head, 32, wr)
    global_store_ptr("py_weakref_head", wr)
    pcc_runtime_log_event_code(4, 1, load_i32(target, 8), 0 if ptr_is_null(callback) != 0 else 1, target)
    return wr


@c_abi_export("py_weakref_call")
def py_weakref_call(ref):
    if ptr_is_null(ref) != 0:
        return null()
    if is_tagged_int(ref) != 0:
        return null()
    if load_i32(ref, 8) != 21:
        return null()
    target = load_ptr(ref, 16)
    if ptr_is_null(target) != 0:
        none_obj = _py_none()
        py_incref(none_obj)
        return none_obj
    resolved = pcc_gc_note_relocation_read(target)
    if ptr_eq(resolved, target) == 0:
        store_ptr(ref, 16, resolved)
        target = resolved
    py_incref(target)
    return target


@c_abi_export("py_weak_value_dict_new")
def py_weak_value_dict_new():
    return py_dict_new()


@c_abi_export("py_weak_value_dict_set")
def py_weak_value_dict_set(d, key, value) -> int:
    if ptr_is_null(d) != 0:
        return -1
    if ptr_is_null(key) != 0:
        return -1
    if ptr_is_null(value) != 0:
        return -1
    wr = py_weakref_new(value, _py_none())
    if ptr_is_null(wr) != 0:
        return -1
    py_dict_set(d, key, wr)
    py_decref(wr)
    return 0


@c_abi_export("py_weak_value_dict_contains")
def py_weak_value_dict_contains(d, key) -> int:
    if ptr_is_null(d) != 0:
        return 0
    if ptr_is_null(key) != 0:
        return 0
    wr = py_dict_get(d, key)
    if ptr_is_null(wr) != 0:
        return 0
    target = py_weakref_call(wr)
    py_decref(wr)
    if ptr_is_null(target) != 0:
        return 0
    if ptr_eq(target, _py_none()) != 0:
        py_decref(target)
        py_dict_del(d, key)
        return 0
    py_decref(target)
    return 1


@c_abi_export("py_weak_value_dict_len")
def py_weak_value_dict_len(d) -> int:
    if ptr_is_null(d) != 0:
        return 0
    keys = py_dict_keys(d)
    if ptr_is_null(keys) != 0:
        return 0
    n: int = py_list_len(keys)
    i: int = 0
    count: int = 0
    while i < n:
        key = py_list_get(keys, i)
        if ptr_is_null(key) == 0:
            if py_weak_value_dict_contains(d, key) != 0:
                count += 1
            py_decref(key)
        i += 1
    py_decref(keys)
    return count


@c_abi_export("py_weak_key_dict_new")
def py_weak_key_dict_new():
    return py_list_new(0)


def _weak_key_entry_new(key, value):
    wr = py_weakref_new(key, _py_none())
    if ptr_is_null(wr) != 0:
        return null()
    entry = py_tuple_new(2)
    if ptr_is_null(entry) != 0:
        py_decref(wr)
        return null()
    py_tuple_set_item(entry, 0, wr)
    py_tuple_set_item(entry, 1, value)
    py_decref(wr)
    return entry


@c_abi_export("py_weak_key_dict_set")
def py_weak_key_dict_set(d, key, value) -> int:
    if ptr_is_null(d) != 0:
        return -1
    if ptr_is_null(key) != 0:
        return -1
    if ptr_is_null(value) != 0:
        return -1
    n: int = py_list_len(d)
    i: int = 0
    while i < n:
        entry = py_list_get(d, i)
        if ptr_is_null(entry) == 0:
            wr = py_tuple_get(entry, 0)
            live = py_weakref_call(wr)
            same: int = 0
            if ptr_is_null(live) == 0:
                if ptr_eq(live, _py_none()) == 0:
                    if ptr_eq(live, key) != 0:
                        same = 1
                py_decref(live)
            if ptr_is_null(wr) == 0:
                py_decref(wr)
            if same != 0:
                replacement = _weak_key_entry_new(key, value)
                if ptr_is_null(replacement) != 0:
                    py_decref(entry)
                    return -1
                py_list_set(d, i, replacement)
                py_decref(replacement)
                py_decref(entry)
                return 0
            py_decref(entry)
        i += 1
    entry2 = _weak_key_entry_new(key, value)
    if ptr_is_null(entry2) != 0:
        return -1
    py_list_append(d, entry2)
    py_decref(entry2)
    return 0


@c_abi_export("py_weak_key_dict_len")
def py_weak_key_dict_len(d) -> int:
    if ptr_is_null(d) != 0:
        return 0
    count: int = 0
    i: int = 0
    while i < py_list_len(d):
        entry = py_list_get(d, i)
        if ptr_is_null(entry) != 0:
            i += 1
        else:
            wr = py_tuple_get(entry, 0)
            live = py_weakref_call(wr)
            is_live: int = 0
            if ptr_is_null(live) == 0:
                if ptr_eq(live, _py_none()) == 0:
                    is_live = 1
                py_decref(live)
            if ptr_is_null(wr) == 0:
                py_decref(wr)
            if is_live != 0:
                count += 1
                i += 1
            else:
                popped = py_list_pop(d, i)
                if ptr_is_null(popped) == 0:
                    py_decref(popped)
            py_decref(entry)
    return count


@c_abi_export("py_weakref_invalidate")
def py_weakref_invalidate(target) -> None:
    if ptr_is_null(target) != 0:
        return
    if is_tagged_int(target) != 0:
        return
    wr = global_load_ptr("py_weakref_head")
    while ptr_is_null(wr) == 0:
        nxt = load_ptr(wr, 40)
        wr_target = load_ptr(wr, 16)
        resolved = pcc_gc_note_relocation_read(wr_target)
        if ptr_eq(wr_target, target) != 0 or ptr_eq(resolved, target) != 0:
            store_ptr(wr, 16, null())
            pcc_runtime_log_event_code(4, 2, 0, load_i32(target, 8), target)
            callback = pcc_gc_load_ptr(wr, ptr_add(wr, 24))
            if ptr_is_null(callback) == 0:
                args = py_tuple_new(1)
                if ptr_is_null(args) == 0:
                    pcc_runtime_log_event_code(4, 3, 0, 0, wr)
                    py_tuple_set_item(args, 0, wr)
                    py_obj_call(callback, args, _py_none())
                    py_decref(args)
                py_clear_exception()
        elif ptr_eq(resolved, wr_target) == 0:
            store_ptr(wr, 16, resolved)
        wr = nxt


@c_abi_export("py_dealloc_weakref")
def py_dealloc_weakref(ref) -> None:
    if ptr_is_null(ref) != 0:
        return
    _unlink(ref)
    pcc_runtime_log_event_code(4, 4, 0, 0, ref)
    callback = pcc_gc_load_ptr(ref, ptr_add(ref, 24))
    if ptr_is_null(callback) == 0:
        py_decref(callback)
    pcc_gc_free_object_memory(ref)

File: pcc/py_runtime/src/py_weakref.c
/* Minimal weakref.ref support for pcc's native Python runtime.
 *
 * Weak references keep a borrowed target pointer and are linked in a
 * process-global list. Object dealloc/in-cycle collection invalidates
 * entries for the target before memory is released.
 */

#include "py_internal.h"
#include <stdlib.h>
#include <string.h>

static PyWeakRefObject *g_weakrefs = NULL;

static void weakref_unlink(PyWeakRefObject *wr) {
    if (wr->prev == wr) {
        wr->prev = NULL;
        wr->next = NULL;
        return;
    }
    if (wr->prev == NULL && g_weakrefs != wr) {
        wr->next = NULL;
        return;
    }
    if (wr->prev != NULL && wr->prev->next != wr) {
        wr->prev = NULL;
        wr->next = NULL;
        return;
    }
    if (wr->prev) wr->prev->next = wr->next;
    else g_weakrefs = wr->next;
    if (wr->next) wr->next->prev = wr->prev;
    wr->prev = NULL;
    wr->next = NULL;
}

PyObject *py_weakref_new(PyObject *target, PyObject *callback) {
    if (target == NULL || PY_IS_TAGGED_INT(target)) return NULL;
    /* Value projections are identity-free; a weak reference observes
     * identity lifetime and a ValueBox's lifetime is unpredictable
     * (every boxing makes a new box). Mirror CPython's analogue
     * (weakref.ref(3) -> TypeError). The compile-time diagnostic
     * catches the static form; this covers Dyn-path boxes. */
    if (py_type_of(target) == PY_TYPE_VALUEBOX) {
        py_raise(py_exc_new(
            PY_EXC_TYPEERROR,
            "cannot create weak reference to a valueclass payload"));
        return NULL;
    }
    if (callback == py_None) callback = NULL;
    target = pcc_gc_note_relocation_read(target);

    PyWeakRefObject *wr = (PyWeakRefObject *)pcc_gc_alloc(
        (int64_t)sizeof(PyWeakRefObject), PY_TYPE_WEAKREF, 0);
    if (wr == NULL) return NULL;
    memset((char *)wr + sizeof(PyObjectHeader), 0,
           sizeof(PyWeakRefObject) - sizeof(PyObjectHeader));
    wr->target = target;
    wr->callback = NULL;
    if (callback != NULL) {
        pcc_gc_store_ptr((PyObject *)wr, &wr->callback, callback);
    }

    wr->next = g_weakrefs;
    if (g_weakrefs != NULL) g_weakrefs->prev = wr;
    g_weakrefs = wr;
    pcc_runtime_log_event("weakref", "new", py_header(target)->type_tag, callback != NULL, target);
    return (PyObject *)wr;
}

PyObject *py_weakref_call(PyObject *ref) {
    if (ref == NULL || PY_IS_TAGGED_INT(ref)) return NULL;
    if (py_header(ref)->type_tag != PY_TYPE_WEAKREF) return NULL;
    PyWeakRefObject *wr = (PyWeakRefObject *)ref;
    if (wr->target == NULL) {
        py_incref(py_None);
        return py_None;
    }
    PyObject *resolved = pcc_gc_note_relocation_read(wr->target);
    if (resolved != wr->target) {
        wr->target = resolved;
    }
    py_incref(wr->target);
    return wr->target;
}

PyObject *py_weak_value_dict_new(void) {
    return py_dict_new();
}

int64_t py_weak_value_dict_set(PyObject *dict, PyObject *key, PyObject *value) {
    if (dict == NULL || key == NULL || value == NULL) return -1;
    PyObject *wr = py_weakref_new(value, py_None);
    if (wr == NULL) return -1;
    py_dict_set(dict, key, wr);
    py_decref(wr);
    return 0;
}

int64_t py_weak_value_dict_contains(PyObject *dict, PyObject *key) {
    if (dict == NULL || key == NULL) return 0;
    PyObject *wr = py_dict_get(dict, key);
    if (wr == NULL) return 0;
    PyObject *target = py_weakref_call(wr);
    py_decref(wr);
    if (target == NULL) return 0;
    if (target == py_None) {
        py_decref(target);
        (void)py_dict_del(dict, key);
        return 0;
    }
    py_decref(target);
    return 1;
}

int64_t py_weak_value_dict_len(PyObject *dict) {
    if (dict == NULL) return 0;
    PyObject *keys = py_dict_keys(dict);
    if (keys == NULL) return 0;
    int64_t n = py_list_len(keys);
    int64_t count = 0;
    for (int64_t i = 0; i < n; i++) {
        PyObject *key = py_list_get(keys, i);
        if (key == NULL) continue;
        if (py_weak_value_dict_contains(dict, key)) count++;
        py_decref(key);
    }
    py_decref(keys);
    return count;
}

PyObject *py_weak_key_dict_new(void) {
    return py_list_new(0);
}

static PyObject *weak_key_entry_new(PyObject *key, PyObject *value) {
    PyObject *wr = py_weakref_new(key, py_None);
    if (wr == NULL) return NULL;
    PyObject *entry = py_tuple_new(2);
    if (entry == NULL) {
        py_decref(wr);
        return NULL;
    }
    py_tuple_set_item(entry, 0, wr);
    py_tuple_set_item(entry, 1, value);
    py_decref(wr);
    return entry;
}

int64_t py_weak_key_dict_set(PyObject *dict, PyObject *key, PyObject *value) {
    if (dict == NULL || key == NULL || value == NULL) return -1;
    int64_t n = py_list_len(dict);
    for (int64_t i = 0; i < n; i++) {
        PyObject *entry = py_list_get(dict, i);
        if (entry == NULL) continue;
        PyObject *wr = py_tuple_get(entry, 0);
        PyObject *live = py_weakref_call(wr);
        int same = live != NULL && live != py_None && live == key;
        if (live != NULL) py_decref(live);
        if (wr != NULL) py_decref(wr);
        if (same) {
            PyObject *replacement = weak_key_entry_new(key, value);
            if (replacement == NULL) {
                py_decref(entry);
                return -1;
            }
            py_list_set(dict, i, replacement);
            py_decref(replacement);
            py_decref(entry);
            return 0;
        }
        py_decref(entry);
    }
    PyObject *entry = weak_key_entry_new(key, value);
    if (entry == NULL) return -1;
    py_list_append(dict, entry);
    py_decref(entry);
    return 0;
}

int64_t py_weak_key_dict_len(PyObject *dict) {
    if (dict == NULL) return 0;
    int64_t count = 0;
    int64_t i = 0;
    while (i < py_list_len(dict)) {
        PyObject *entry = py_list_get(dict, i);
        if (entry == NULL) {
            i++;
            continue;
        }
        PyObject *wr = py_tuple_get(entry, 0);
        PyObject *live = py_weakref_call(wr);
        int is_live = live != NULL && live != py_None;
        if (live != NULL) py_decref(live);
        if (wr != NULL) py_decref(wr);
        if (is_live) {
            count++;
            i++;
        } else {
            PyObject *popped = py_list_pop(dict, i);
            if (popped != NULL) py_decref(popped);
        }
        py_decref(entry);
    }
    return count;
}

void py_weakref_invalidate(PyObject *target) {
    if (target == NULL || PY_IS_TAGGED_INT(target)) return;
    PyWeakRefObject *wr = g_weakrefs;
    while (wr != NULL) {
        PyWeakRefObject *next = wr->next;
        PyObject *wr_target = wr->target;
        PyObject *resolved = pcc_gc_note_relocation_read(wr_target);
        if (wr_target == target || resolved == target) {
            wr->target = NULL;
            pcc_runtime_log_event("weakref", "invalidate", 0, py_header(target)->type_tag, target);
            PyObject *callback = pcc_gc_load_ptr((PyObject *)wr, &wr->callback);
            if (callback != NULL) {
                PyObject *args = py_tuple_new(1);
                if (args != NULL) {
                    pcc_runtime_log_event("weakref", "callback", 0, 0, (PyObject *)wr);
                    py_tuple_set_item(args, 0, (PyObject *)wr);
                    /* Weakref callbacks are an unraisable boundary: the
                     * callback result is ignored, but its owned reference
                     * must still be released.  py_obj_call classifies a
                     * silent NULL as RuntimeError before we deliberately
                     * clear the unraisable exception below. */
                    PyObject *result = py_obj_call(callback, args, py_None);
                    if (result != NULL) py_decref(result);
                    py_decref(args);
                }
                py_clear_exception();
            }
        } else if (resolved != wr_target) {
            wr->target = resolved;
        }
        wr = next;
    }
}

int64_t py_weakref_retarget(PyObject *from_obj, PyObject *to_obj) {
    if (from_obj == NULL || to_obj == NULL) return -1;
    if (PY_IS_TAGGED_INT(from_obj) || PY_IS_TAGGED_INT(to_obj)) return -1;
    if (
        py_header(from_obj)->type_tag != PY_TYPE_WEAKREF
        || py_header(to_obj)->type_tag != PY_TYPE_WEAKREF
    ) return -1;

    PyWeakRefObject *from = (PyWeakRefObject *)from_obj;
    PyWeakRefObject *to = (PyWeakRefObject *)to_obj;
    int found = 0;
    for (PyWeakRefObject *wr = g_weakrefs; wr != NULL; wr = wr->next) {
        if (wr == from) {
            found = 1;
            break;
        }
    }
    if (!found) return -1;

    to->prev = from->prev;
    to->next = from->next;
    if (from->prev != NULL) {
        from->prev->next = to;
    } else {
        g_weakrefs = to;
    }
    if (from->next != NULL) {
        from->next->prev = to;
    }

    from->prev = from;
    from->next = NULL;
    return 0;
}

void py_dealloc_weakref(PyObject *ref) {
    if (ref == NULL) return;
    PyWeakRefObject *wr = (PyWeakRefObject *)ref;
    weakref_unlink(wr);
    pcc_runtime_log_event("weakref", "dealloc", 0, 0, ref);
    PyObject *callback = pcc_gc_load_ptr(ref, &wr->callback);
    if (callback != NULL) py_decref(callback);
    pcc_gc_free_object_memory(ref);
}
